man/cyrus.conf.5
.\" -*- nroff -*-
.TH CYRUS.CONF 5 "Project Cyrus" CMU
.\"
.\" Copyright (c) 1994-2008 Carnegie Mellon University.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\"
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in
.\"    the documentation and/or other materials provided with the
.\"    distribution.
.\"
.\" 3. The name "Carnegie Mellon University" must not be used to
.\"    endorse or promote products derived from this software without
.\"    prior written permission. For permission or any legal
.\"    details, please contact
.\"      Carnegie Mellon University
.\"      Center for Technology Transfer and Enterprise Creation
.\"      4615 Forbes Avenue
.\"      Suite 302
.\"      Pittsburgh, PA  15213
.\"      (412) 268-7393, fax: (412) 268-7395
.\"      innovation@andrew.cmu.edu
.\"
.\" 4. Redistributions of any form whatsoever must retain the following
.\"    acknowledgment:
.\"    "This product includes software developed by Computing Services
.\"     at Carnegie Mellon University (http://www.cmu.edu/computing/)."
.\"
.\" CARNEGIE MELLON UNIVERSITY DISCLAIMS ALL WARRANTIES WITH REGARD TO
.\" THIS SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
.\" AND FITNESS, IN NO EVENT SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE
.\" FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
.\" AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.SH NAME
cyrus.conf \- Cyrus configuration file
.SH DESCRIPTION
\fB/etc/cyrus.conf\fR
is the configuration file for the Cyrus \fBmaster\fR process.  It
defines the startup procedures, services and events to be spawned by
\fBmaster\fR.
.PP
The \fB/etc/cyrus.conf\fR file consists of a series of entries divided
into sections of the form
.P
.RS
\fIsection\fR {
.RS
\fIname arguments
.br
        ...
.br
        ...
.br
        ...
\fR
.RE
}
.RE
.PP
where \fIsection\fR is the name of the section, \fIname\fR is the name
of the entry and \fIarguments\fR is the whitespace-separated list of
arguments for the entry.  The \fIname\fR may be any sequence of
alphabetic and numeric characters, but may not contain punctuation
such as '-' or '_'.  In the \fBSERVICES\fP section, names must be
unique.
.PP
Blank lines and lines beginning with ``#'' are ignored.
.SH SECTION DESCRIPTIONS
The paragraphs below detail the three sections (\fBSTART\fR,
\fBSERVICES\fR, \fBEVENTS\fR) that can be placed in the
\fB/etc/cyrus.conf\fR file.  The arguments that are available for each
entry within the section are described, and each argument's default
value is shown.
.PP
Arguments can appear in any order.
Some arguments have no default value, these are listed with
``<no default>''.  For string arguments, the value MUST be enclosed in
double quotes.
.SS START
This section lists the processes to run before any
\fBSERVICES\fR are spawned.  This section is typically used to
initialize databases and start long running daemons.
.IP "\fBcmd=\fR<no default>" 5
The command (with options) to spawn as a child process.  This string argument
is required.
.SS SERVICES
This section is the heart of the \fB/etc/cyrus.conf\fR file.  It lists
the processes that should be spawned to handle client connections made
on certain Internet/UNIX sockets.
.IP "\fBbabysit=\fR0" 5
Integer value - if non-zero, will make sure at least one process is
pre-forked, and will set the maxforkrate to 10 if it's zero.
.IP "\fBcmd=\fR<no default>" 5
The command (with options) to spawn as a child process.  This string
argument is required.
.IP "\fBlisten=\fR<no default>" 5
The UNIX or internet socket to listen on.  This
string field is required and takes one of the following forms:

\fIpath\fR
.br
\fR[ \fIhost\fR \fB: \fR] \fIport\fR
.br
.sp
where \fIpath\fR is the explicit path to a UNIX socket, \fIhost\fR is
either the hostname or bracket-enclosed IP address of a network
interface, and \fIport\fR is either a port number or service name (as listed
in \fB/etc/services\fR).
.IP "\fBproto=\fRtcp" 5
The protocol used for this service (\fItcp\fR, \fItcp4\fR, \fItcp6\fR,
\fIudp\fR, \fIudp4\fR, \fIudp6\fR).  This string argument is optional.

\fBtcp4\fR, \fBudp4\fR: These arguments are used to bind the service to IPv4
only.
.br
\fBtcp6\fR, \fBudp6\fR: These arguments are used to bind the service to IPv6
only, if the operating system supports this.
.br
\fBtcp\fR, \fBudp\fR: These arguments are used to bind to both IPv4 and IPv6
if possible.
.IP "\fBprefork=\fR0" 5
The number of instances of this service to always have running and
waiting for a connection (for faster initial response time).  This
integer value is optional.  Note that if you are listening on multiple
network types (i.e. ipv4 and ipv6) then one process will be forked for
each address, causing twice as many processes as you might expect.
.IP "\fBmaxchild=\fR-1" 5
The maximum number of instances of this service to spawn.  A value of
-1 means unlimited.  This integer value is optional.
.IP "\fBmaxfds=\fR256" 5
The maximum number of file descriptors to which to limit this process.
This integer value is optional.
.IP "\fBmaxforkrate=\fR0" 5
Maximum number of processes to fork per second - the master will insert
sleeps to ensure it doesn't fork faster than this on average.
.SS EVENTS
This section lists processes that should be run at specific intervals,
similar to cron jobs.  This section is typically used to perform
scheduled cleanup/maintenance.
.IP "\fBcmd=\fR<no default>" 5
The command (with options) to spawn as a child process.  This string
argument is required.
.IP "\fBperiod=\fR0" 5
The interval (in minutes) at which to run the command.  This integer value is
optional, but SHOULD be a positive integer > 10.
.IP "\fBat=\fR<hhmm>" 5
The time (24-hour format) at which to run the command each day.  If
set to a valid time (0000-2359), period is automatically set to 1440.
This string argument is optional.
.SH EXAMPLE
# example cyrus.conf

START {
  recover       cmd="ctl_cyrusdb -r"
.br
}

SERVICES {
  imap          cmd="imapd" listen="imap" prefork=1
  imaps         cmd="imapd -s" listen="imaps" prefork=0
  lmtpunix      cmd="lmtpd" listen="/var/imap/socket/lmtp"
  lmtp          cmd="lmtpd" listen="localhost:lmtp"
.br
}

EVENTS {
  checkpoint    cmd="ctl_cyrusdb -c" period=30
  delprune      cmd="cyr_expire -E 3" at=0400
  tlsprune      cmd="tls_prune" at=0400
.br
}
.SH ACCESS CONTROL
When TCP Wrappers is used to control access to Cyrus services, the
\fIname\fR of the service entry should be used as the process name in
the \fBhosts_access(5)\fR table.  For instance, in the example above,
"imap", "imaps", "lmtpunix" and "lmtp" would be used as the process
names.  This allows a single daemon such as imapd to be run in
different modes or configurations (i.e., SSL and non-SSL enabled) yet
still have separate access control rules.
.SH SEE ALSO
.PP
\fBmaster(8)\fR, \fBimapd(8)\fR, \fBpop3d(8)\fR, \fBlmtpd(8)\fR,
\fBtimsieved(8)\fR, \fBidled(8)\fR, \fBnotifyd(8)\fR, \fBctl_cyrusdb(8)\fR,
\fBctl_deliver(8)\fR, \fBtls_prune(8)\fR, \fBhosts_access(5)\fR


sieve/NEWS
CMU Sieve 2.1
-------------

- Compliant with RFC 3028.  As a result, fileinto and redirect only
  accept a single string and NOT a string-list.

- Compliant with draft-martin-sieve-notify-01.  As a result, notify
  actions will need to be updated to the new syntax.

CMU Sieve 2.0
-------------

- Compliant with draft-showalter-sieve-11.txt and
  draft-showalter-sieve-vacation-03.txt.

- Added support for the regex, imapflags, notify and subaddress extensions.
  See README for references.

- Verifies email addresses in redirect and vacation actions are syntactically
  correct (compliant with RFC822).

- Run-time error reporting.

- Changed callback interface to use callback contexts instead of individual
  parameters.  Also added an error string buffer for run-time error reporting.

- Vacation will not reply to any message containing an "auto-submitted"
  header containing anything other than "no".

CMU Sieve 1.4
-------------

Now included with imapd distribution (hell, why not?).

Error returning and recovering:
        added error recovering to the parser (but not much!)
        added error messages to the parser

Working on error returning and error recovering.
        run-time errors
        detect some errors in lexer?

Working on even better parsing:
        verify addresses could be addresses
        verify mailboxes could be mailboxes
        verify outgoing headers can be headers

CMU Sieve 1.3
-------------

Changed for integration with cyrus deliver.

CMU Sieve 1.2
-------------

Added additional callbacks (ok, so I want to make my integration with deliver
easier) and envelope and vacation support.

Made it compile without libcyrus.
It should compile without libcyrus, but then it does not implement the
"address" test.  That's just too much work to do when I have a neato
library to do it for me.

Todo:
- regex matching

CMU Sieve 1.1
-------------

- Updated to draft-showalter-sieve-07bis.txt

- Simple API (see sieve_interface.h; currently mostly undocumented)

- Implements all of the optional features except "envelope"

- Maintains "if it parses, it probably runs" behavior. (Goal: minimize
  run-time errors.)

CMU Sieve 1.0
-------------

- prototype implementation


contrib/cyrus-graphtools.1.0/script/cyrus.pl
#!/usr/bin/perl

# This will read information from the cyrus MIB for all devices specified
# in cyrusrc
#
# Copyright (c) 1994-2008 Carnegie Mellon University.  All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in
#    the documentation and/or other materials provided with the
#    distribution.
#
# 3. The name "Carnegie Mellon University" must not be used to
#    endorse or promote products derived from this software without
#    prior written permission. For permission or any legal
#    details, please contact
#      Carnegie Mellon University
#      Center for Technology Transfer and Enterprise Creation
#      4615 Forbes Avenue
#      Suite 302
#      Pittsburgh, PA  15213
#      (412) 268-7393, fax: (412) 268-7395
#      innovation@andrew.cmu.edu
#
# 4. Redistributions of any form whatsoever must retain the following
#    acknowledgment:
#    "This product includes software developed by Computing Services
#     at Carnegie Mellon University (http://www.cmu.edu/computing/)."
#
# CARNEGIE MELLON UNIVERSITY DISCLAIMS ALL WARRANTIES WITH REGARD TO
# THIS SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
# AND FITNESS, IN NO EVENT SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE
# FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
# AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
# OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
#
# Author: Alison Greenwald <alison@andrew.cmu.edu>

use RRDs;
use SNMP 1.8;

do "/data/prog/cyrus/cyrusrc";
get_data();

sub get_data{
  foreach $hname (keys %HOSTS){
    $MAX = 0;
    %walk=snmp_walk($hname, $HOSTS{$hname}, $MASTER);
    foreach $OID (sort keys %walk){
      @O = split /\./, $OID;
      $RVAL=$walk{$OID};
      chomp $RVAL;
      if($O[-1] > $MAX){
        $MAX = $O[-1];
      }
      $STUFF{"$O[$#O-1]-$O[$#O]"} = $RVAL;
    } #foreach oid

    for($i=1; $i<=$MAX; $i++){
      $blah[0]=$STUFF{"3-$i"};
      $blah[1]=$STUFF{"2-$i"};
      $blah[2]=$STUFF{"1-$i"};
      print "$hname:$blah[0]: $blah[2], $blah[1]\n";
      populate_dbs();
    } #for
  } #foreach hname
} #sub

sub populate_dbs{
  if(open(DB, "< $DPTH/$hname:$blah[0].rrd")){
    close(DB);
  } else {
    print("Making $DPTH/$hname:$blah[0]:daily.rrd\n");
    RRDs::create("$DPTH/$hname:$blah[0].rrd","-s 1",
                 "DS:current:GAUGE:300:U:U",
                 "DS:total:COUNTER:300:U:U",
                 "RRA:MAX:0.5:300:4320",
                 "RRA:MAX:0.5:1800:336",
                 "RRA:MAX:0.5:14400:168",
                 "RRA:MAX:0.5:86400:364");
    $ERROR=RRDs::error;
    print $ERROR if $ERROR;
  }

  RRDs::update("$DPTH/$hname:$blah[0].rrd", "N:$blah[1]:$blah[2]");

}#find_dbs

sub snmp_walk{
  my ($server, $comm, $rootoid) = @_;
  my %walk=();
  my $sess = new SNMP::Session ( DestHost => $server,
                                 Community => $comm,
                                 UseNumeric => 1,
                                 UseLongNames => 1
                               );

  my @orig=split /\./, $rootoid;  # original oid for comparison

  my $var = new SNMP::Varbind(["$rootoid"]);
  my $val = $sess->getnext($var);
  my $name = $var->[$SNMP::Varbind::tag_f];
  $name .= ".$var->[$SNMP::Varbind::iid_f]" if $var->[$SNMP::Varbind::iid_f];
  my @current=split /\./, $name;

  while (!$sess->{ErrorStr} && $orig[$#orig] eq $current[$#orig]
        && $#current > $#orig){
    my $value=$var->[$SNMP::Varbind::val_f];

    $walk{"$name"} = $value;
    $val = $sess->getnext($var);
    $name=$var->[$SNMP::Varbind::tag_f];
    $name .= ".$var->[$SNMP::Varbind::iid_f]" if $var->[$SNMP::Varbind::iid_f];
    @current=split /\./, $name;
  }  #while

  print("$sess->{ErrorStr}\n") if $sess->{ErrorStr};
  return(%walk);

}
